# Google sheet with personal questions

https://docs.google.com/spreadsheets/d/1kH6AfE3Iu6kdgQsDQvEGXaAkKlhAT6VaxeFOlJpnC0E/edit?usp=sharing

Every column corresponds to a single question, every row to a single student.

For example, R. Daneel Olivaw need to report questions 1.1, 1.2, 2.2, 2.3, etc.




# Submiting the answers


Google form to submit your answers: https://docs.google.com/forms/d/e/1FAIpQLSd2gNxCq9j8ZE_NzZlV8Q-gTNbiic5Rkx3vitz7vMlfzkzAzQ/viewform?usp=sf_link

Google form has fields for all questions, but you only need to answer **your** questions (from google sheet above).

Use your **skoltech email**. For Name, Surname use **exactly the same spelling** as in canvas system.

---

Every question has an information about the type of the answer, e.g.

> Observe top 10 observations (int)

here your answer must be a single **integer** number.

---

If your answer is a ``float number``, then it must be provided with **3 decimals after the floating point**, e.g. 1.234

---

If your answer is a ``list of float or integer numbers or str``, then they should be reported in descending (alphabetical) order, without spacing, divided by a comma, e.g.:

10.453,9.112,5.001,5.000 - Right

10.453, 9.112, 5.001, 5.000 - WRONG

---

Part of the tasks, e.g. Q19.1-5, Q20.1-5 do not have corresponding fields in the google form. They are **not optional** and they will be graded manually from your .ipynb file.

---

If you have any questions regarding this Home Assignment, ask them piazza topic: https://piazza.com/class/kespugtqfrn12g?cid=12

# Assignment 1. House Pricing.
by Anvar Kurmukov

---

By the end of this task you will be able to manipulate huge tabular data:
1. Compute different column's statistics (min, max, mean, quantiles etc.);
2. Select observations/features by condition/index;
3. Create new non-linear combinations of the columns (feature engineering);
4. Perform automated data cleaning;

and more.

---

For those who are not familiar with `pandas` we recommend these (alternative) tutorials:

1. Single notebook, covers basic pandas functionality (starting with renaming columns ending with using map, apply etc) ~ 30 short examples with links on videos https://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb . Highly recommended for everyone. (about 1-3 hours to go through)

2. https://github.com/guipsamora/pandas_exercises/ 11 topics covering all essential functionality with excersises (with solutions).

This task will be an easy ride after these tutorials.

---

We are using house sale price data from King County, Wahington, USA. This dataset is in public domain and can be obtained from Kaggle: https://www.kaggle.com/harlfoxem/housesalesprediction

You need to place `kc_nouse_data.csv` file in the same directory as this notebook.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# 1. Loading data

As always in Data Science you are starting with making nice cup of tea (or coffee). Your next move is to load the data:

- Start with loading `house_data.csv` file using `pd.read_csv()` function.
- You may also want to increase maximal displayed pandas columns: set `pd.options.display.max_columns` to 30
- Print top 10 observations in the table. `.head()`
- Print last 10 observations in the table. `.tail()`
- Print all the data columns names using method `.columns`
- Print data size (number of rows and columns). This is the `.shape` of the data.

*Almost* every python has a `head` and a `tail` just as DataFrames do.

In [2]:
# Load the data
house_data = pd.read_csv('kc_house_data.csv')


In [3]:
# Observe top 10 observations (int)

# Q1.1 What is the price of a house with `id` == 7237550310?
# +Q1.2 How many bedrooms has a house with `id` == 7237550310?
# Q1.3 When was the house with `id` == 2414600126 built (`yr_built`)?
# +Q1.4 What is the `grade` of a house with `id` == 5631500400?
# Q1.5 When was the house with `id` == 6414100192 renovated (`yr_renovated`)?
 
house_data.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0         3       1.00         1180   
1  6414100192  20141209T000000   538000.0         3       2.25         2570   
2  5631500400  20150225T000000   180000.0         2       1.00          770   
3  2487200875  20141209T000000   604000.0         4       3.00         1960   
4  1954400510  20150218T000000   510000.0         3       2.00         1680   
5  7237550310  20140512T000000  1225000.0         4       4.50         5420   
6  1321400060  20140627T000000   257500.0         3       2.25         1715   
7  2008000270  20150115T000000   291850.0         3       1.50         1060   
8  2414600126  20150415T000000   229500.0         3       1.00         1780   
9  3793500160  20150312T000000   323000.0         3       2.50         1890   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   
5    101930     1.0           0     0  ...     11        3890           1530   
6      6819     2.0           0     0  ...      7        1715              0   
7      9711     1.0           0     0  ...      7        1060              0   
8      7470     1.0           0     0  ...      7        1050            730   
9      6560     2.0           0     0  ...      7        1890              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   
5      2001             0    98053  47.6561 -122.005           4760   
6      1995             0    98003  47.3097 -122.327           2238   
7      1963             0    98198  47.4095 -122.315           1650   
8      1960             0    98146  47.5123 -122.337           1780   
9      2003             0    98038  47.3684 -122.031           2390   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  
5      101930  
6        6819  
7        9711  
8        8113  
9        7570  

[10 rows x 21 columns]

In [4]:
# +Q1.2 How many bedrooms has a house with `id` == 7237550310?

house_data[house_data['id'] == 7237550310] # 4 bedrooms

id             date      price  bedrooms  bathrooms  sqft_living  \
5  7237550310  20140512T000000  1225000.0         4        4.5         5420   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
5    101930     1.0           0     0  ...     11        3890           1530   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
5      2001             0    98053  47.6561 -122.005           4760   

   sqft_lot15  
5      101930  

[1 rows x 21 columns]

In [5]:
# +Q1.4 What is the `grade` of a house with `id` == 5631500400?

house_data[house_data['id'] == 5631500400] # grade 6

id             date     price  bedrooms  bathrooms  sqft_living  \
2  5631500400  20150225T000000  180000.0         2        1.0          770   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
2     10000     1.0           0     0  ...      6         770              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
2      1933             0    98028  47.7379 -122.233           2720   

   sqft_lot15  
2        8062  

[1 rows x 21 columns]

In [6]:
# Observe last 10 observations (int)

# +Q2.1 What is the price of a house with `id` == 263000018?
# +Q2.2 How many bedrooms has a house with `id` == 291310100?
# Q2.3 When was the house with `id` == 1523300141 built (`yr_built`)?
# Q2.4 How many floors house with `id` == 2997800021 has?
# Q2.5 What is the zipcode of the house with `id` == 7852140040?

house_data.tail(10)

id             date      price  bedrooms  bathrooms  \
21603  7852140040  20140825T000000   507250.0         3       2.50   
21604  9834201367  20150126T000000   429000.0         3       2.00   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1007500.0         4       3.50   
21607  2997800021  20150219T000000   475000.0         3       2.50   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21603         2270      5536     2.0           0     0  ...      8   
21604         1490      1126     3.0           0     0  ...      8   
21605         2520      6023     2.0           0     0  ...      9   
21606         3510      7200     2.0           0     0  ...      9   
21607         1310      1294     2.0           0     0  ...      8   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21603        2270              0      2003             0    98065  47.5389   
21604        1490              0      2014             0    98144  47.5699   
21605        2520              0      2014             0    98056  47.5137   
21606        2600            910      2009             0    98136  47.5537   
21607        1180            130      2008             0    98116  47.5773   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21603 -121.881           2270        5731  
21604 -122.288           1400        1230  
21605 -122.167           2520        6023  
21606 -122.398           2050        6200  
21607 -122.409           1330        1265  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[10 rows x 21 columns]

In [7]:
# +Q2.1 What is the price of a house with `id` == 263000018?
house_data[house_data['id'] == 263000018] # price 360000.0

id             date     price  bedrooms  bathrooms  sqft_living  \
21608  263000018  20140521T000000  360000.0         3        2.5         1530   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
21608      1131     3.0           0     0  ...      8        1530   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
21608              0      2009             0    98103  47.6993 -122.346   

       sqft_living15  sqft_lot15  
21608           1530        1509  

[1 rows x 21 columns]

In [8]:
# +Q2.2 How many bedrooms has a house with `id` == 291310100?
house_data[house_data['id'] == 291310100] # 3 bedrooms

id             date     price  bedrooms  bathrooms  sqft_living  \
21611  291310100  20150116T000000  400000.0         3        2.5         1600   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
21611      2388     2.0           0     0  ...      8        1600   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
21611              0      2004             0    98027  47.5345 -122.069   

       sqft_living15  sqft_lot15  
21611           1410        1287  

[1 rows x 21 columns]

In [9]:
house_data[house_data['bedrooms'] == 0]

id             date      price  bedrooms  bathrooms  \
875    6306400140  20140612T000000  1095000.0         0       0.00   
3119   3918400017  20150205T000000   380000.0         0       0.00   
3467   1453602309  20140805T000000   288000.0         0       1.50   
4868   6896300380  20141002T000000   228000.0         0       1.00   
6994   2954400190  20140624T000000  1295650.0         0       0.00   
8477   2569500210  20141117T000000   339950.0         0       2.50   
8484   2310060040  20140925T000000   240000.0         0       2.50   
9773   3374500520  20150429T000000   355000.0         0       0.00   
9854   7849202190  20141223T000000   235000.0         0       0.00   
12653  7849202299  20150218T000000   320000.0         0       2.50   
14423  9543000205  20150413T000000   139950.0         0       0.00   
18379  1222029077  20141029T000000   265000.0         0       0.75   
19452  3980300371  20140926T000000   142000.0         0       0.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
875           3064      4764     3.5           0     2  ...      7   
3119          1470       979     3.0           0     2  ...      8   
3467          1430      1650     3.0           0     0  ...      7   
4868           390      5900     1.0           0     0  ...      4   
6994          4810     28008     2.0           0     0  ...     12   
8477          2290      8319     2.0           0     0  ...      8   
8484          1810      5669     2.0           0     0  ...      7   
9773          2460      8049     2.0           0     0  ...      8   
9854          1470      4800     2.0           0     0  ...      7   
12653         1490      7111     2.0           0     0  ...      7   
14423          844      4269     1.0           0     0  ...      7   
18379          384    213444     1.0           0     0  ...      4   
19452          290     20875     1.0           0     0  ...      1   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
875          3064              0      1990             0    98102  47.6362   
3119         1470              0      2006             0    98133  47.7145   
3467         1430              0      1999             0    98125  47.7222   
4868          390              0      1953             0    98118  47.5260   
6994         4810              0      1990             0    98053  47.6642   
8477         2290              0      1985             0    98042  47.3473   
8484         1810              0      2003             0    98038  47.3493   
9773         2460              0      1990             0    98031  47.4095   
9854         1470              0      1996             0    98065  47.5265   
12653        1490              0      1999             0    98065  47.5261   
14423         844              0      1913             0    98001  47.2781   
18379         384              0      2003             0    98070  47.4177   
19452         290              0      1963             0    98024  47.5308   

          long  sqft_living15  sqft_lot15  
875   -122.322           2360        4000  
3119  -122.356           1470        1399  
3467  -122.290           1430        1650  
4868  -122.261           2170        6000  
6994  -122.069           4740       35061  
8477  -122.151           2500        8751  
8484  -122.053           1810        5685  
9773  -122.168           2520        8050  
9854  -121.828           1060        7200  
12653 -121.826           1500        4675  
14423 -122.250           1380        9600  
18379 -122.491           1920      224341  
19452 -121.888           1620       22850  

[13 rows x 21 columns]

In [10]:
# Increase maximal displayed columns
pd.options.display.max_columns = 30
pd.options.display.max_columns

30

In [11]:
# Observe top 10 observations again
# Is there any new columns displayed?
house_data.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0         3       1.00         1180   
1  6414100192  20141209T000000   538000.0         3       2.25         2570   
2  5631500400  20150225T000000   180000.0         2       1.00          770   
3  2487200875  20141209T000000   604000.0         4       3.00         1960   
4  1954400510  20150218T000000   510000.0         3       2.00         1680   
5  7237550310  20140512T000000  1225000.0         4       4.50         5420   
6  1321400060  20140627T000000   257500.0         3       2.25         1715   
7  2008000270  20150115T000000   291850.0         3       1.50         1060   
8  2414600126  20150415T000000   229500.0         3       1.00         1780   
9  3793500160  20150312T000000   323000.0         3       2.50         1890   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0           0     0          3      7        1180   
1      7242     2.0           0     0          3      7        2170   
2     10000     1.0           0     0          3      6         770   
3      5000     1.0           0     0          5      7        1050   
4      8080     1.0           0     0          3      8        1680   
5    101930     1.0           0     0          3     11        3890   
6      6819     2.0           0     0          3      7        1715   
7      9711     1.0           0     0          3      7        1060   
8      7470     1.0           0     0          3      7        1050   
9      6560     2.0           0     0          3      7        1890   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   
5           1530      2001             0    98053  47.6561 -122.005   
6              0      1995             0    98003  47.3097 -122.327   
7              0      1963             0    98198  47.4095 -122.315   
8            730      1960             0    98146  47.5123 -122.337   
9              0      2003             0    98038  47.3684 -122.031   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  
5           4760      101930  
6           2238        6819  
7           1650        9711  
8           1780        8113  
9           2390        7570

Ans: yes, there is one new column - condition, where was '...' before

In [12]:
# Print all the columns/features names (int)

# +Q3.1 How many columns have `yr_` prefix?
# Q3.2 How many columns have `sqft_` prefix?
# Q3.3 How many columns assosiated with house earth coordinates are in the data?
# +Q3.4 How many columns have `rooms` in their names?
house_data.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

Ans: 3.1) 2 columns;
3.4) 2 columns.

In [13]:
# Print data size (int)

# Q4.1 How many observations are in the data?
# Q4.2 How many features are in the data?

house_data.shape

(21613, 21)

Ans: 21613 observations and 21 features

# 2. Basic data exploration

Lets do some basics:
- `.count()` number of not NaN's in every column.
- Is there any missing values in the data?
- Count number of unique values in every column `.nunique()`. 
- What does this tells you about the features, which are most likely categorical and which are most likely numerical?
- Use pandas `.describe()` to display basic statistic about the data.
- Use pandas `.value_counts()` to count number of unique values in a specific column.
- Use pandas `.min()`, `.max()`, `.mean()`, `.std()` to display specific statistics about the data.
- Use pandas `.dtypes` field to display data types in columns.

**Hint**
You could use `.sort_index()` or `.sort_values()` to sort the result of `.value_counts()`

In [14]:
# Display number of not NaN's in every column (int)

# Q5.1 How many NA values are in the `floors` column?
# +Q5.2 How many NA values are in the `grade` column? 0
# +Q5.3 How many NA values are in the `bedrooms` column? 0
# Q5.4 How many NA values are in the `yr_built` column?
# Q5.5 How many NA values (not zeros, but empty, missing values) are in the `yr_renovated` column?

house_data.count()

id               21613
date             21613
price            21613
bedrooms         21613
bathrooms        21613
sqft_living      21613
sqft_lot         21613
floors           21613
waterfront       21613
view             21613
condition        21613
grade            21613
sqft_above       21613
sqft_basement    21613
yr_built         21613
yr_renovated     21613
zipcode          21613
lat              21613
long             21613
sqft_living15    21613
sqft_lot15       21613
dtype: int64

In [15]:
house_data.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [16]:
house_data['bedrooms'][house_data['bedrooms'].isna()]

Series([], Name: bedrooms, dtype: int64)

In [17]:
house_data.isna()

id   date  price  bedrooms  bathrooms  sqft_living  sqft_lot  \
0      False  False  False     False      False        False     False   
1      False  False  False     False      False        False     False   
2      False  False  False     False      False        False     False   
3      False  False  False     False      False        False     False   
4      False  False  False     False      False        False     False   
...      ...    ...    ...       ...        ...          ...       ...   
21608  False  False  False     False      False        False     False   
21609  False  False  False     False      False        False     False   
21610  False  False  False     False      False        False     False   
21611  False  False  False     False      False        False     False   
21612  False  False  False     False      False        False     False   

       floors  waterfront   view  condition  grade  sqft_above  sqft_basement  \
0       False       False  False      False  False       False          False   
1       False       False  False      False  False       False          False   
2       False       False  False      False  False       False          False   
3       False       False  False      False  False       False          False   
4       False       False  False      False  False       False          False   
...       ...         ...    ...        ...    ...         ...            ...   
21608   False       False  False      False  False       False          False   
21609   False       False  False      False  False       False          False   
21610   False       False  False      False  False       False          False   
21611   False       False  False      False  False       False          False   
21612   False       False  False      False  False       False          False   

       yr_built  yr_renovated  zipcode    lat   long  sqft_living15  \
0         False         False    False  False  False          False   
1         False         False    False  False  False          False   
2         False         False    False  False  False          False   
3         False         False    False  False  False          False   
4         False         False    False  False  False          False   
...         ...           ...      ...    ...    ...            ...   
21608     False         False    False  False  False          False   
21609     False         False    False  False  False          False   
21610     False         False    False  False  False          False   
21611     False         False    False  False  False          False   
21612     False         False    False  False  False          False   

       sqft_lot15  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
21608       False  
21609       False  
21610       False  
21611       False  
21612       False  

[21613 rows x 21 columns]

In [18]:
# Count number of unique values in every column (int)

# +Q6.1 How many unique values are in the `bedrooms` column? 13
# +Q6.2 How many unique values are in the `grade` column? 12
# Q6.3 How many unique values are in the `yr_renovated` column?
# Q6.4 How many unique values are in the `bathrooms` column?
# Q6.5 How many unique values are in the `long` column?

house_data.nunique()

id               21436
date               372
price             4028
bedrooms            13
bathrooms           30
sqft_living       1038
sqft_lot          9782
floors               6
waterfront           2
view                 5
condition            5
grade               12
sqft_above         946
sqft_basement      306
yr_built           116
yr_renovated        70
zipcode             70
lat               5034
long               752
sqft_living15      777
sqft_lot15        8689
dtype: int64

What does this tells you about the features, which are most likely categorical and which are most likely numerical?

Ans: columns with only few unique values are more likely categorical, while columns with a lot of different values - numerical.

In [19]:
# Count frequency of the values in different columns (list of ints in descending order)
# You could select a column using same syntax as for selecting a key from a dictionary: `data[colname]`

# +Q7.1 For every unique `floors` value give its number of occurences.
# +Q7.2 For every unique `condition` value give its number of occurences.
# Q7.3 For every unique `bedrooms` value give its number of occurences.
# Q7.4 For every unique `grade` value give its number of occurences.
# Q7.5 For every unique `view` value give its number of occurences.

house_data['floors'].unique(), house_data['condition'].unique()

(array([1. , 2. , 1.5, 3. , 2.5, 3.5]), array([3, 5, 4, 1, 2], dtype=int64))

In [20]:
house_data['floors'].value_counts().tolist()

[10680, 8241, 1910, 613, 161, 8]

In [21]:
house_data['condition'].value_counts().tolist()

[14031, 5679, 1701, 172, 30]

In [22]:
# Display basic data statistics using .describe()
house_data.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [23]:
# Display some column statistics (list of floats, rounded up to 3 digits, e.g. 1.234)

# +Q8.1 What are the max, min, mean and the std of the `floors` column?
# Q8.2 What are the max, min, mean and the std of the `bedrooms` column?
# +Q8.3 What are the max, min, mean and the std of the `sqft_living` column?
# Q8.4 What are the max, min, mean and the std of the `price` column?
# Q8.5 What are the max, min, mean and the std of the `long` column?

a = house_data['floors']
a_fl = [round(a.max(),3), round(a.min(),3), round(a.mean(),3), round(a.std(),3)]
a_fl

[3.5, 1.0, 1.494, 0.54]

In [24]:
a = house_data['sqft_living']
a_sq_liv = [round(a.max(),3), round(a.min(),3), round(a.mean(),3), round(a.std(),3)]
a_sq_liv

[13540, 290, 2079.9, 918.441]

In [25]:
# Display data types of all columns (int)

# +Q9.1 How many columns have `object` data type? 1
# Q9.2 How many columns have `int64` data type?
# Q9.3 How many columns have `float64` data type?

# Display data types of all columns (list of str)
# Q9.4 What are the columns with dtype == `float64`?
# +Q9.5 What are the columns with dtype == `int64`? 
house_data.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [26]:
a_int64 = house_data.dtypes[house_data.dtypes == 'int64'].index.tolist()  # 9.5
a_int64.sort()
a_int64

['bedrooms',
 'condition',
 'grade',
 'id',
 'sqft_above',
 'sqft_basement',
 'sqft_living',
 'sqft_living15',
 'sqft_lot',
 'sqft_lot15',
 'view',
 'waterfront',
 'yr_built',
 'yr_renovated',
 'zipcode']

bedrooms,condition,grade,id,sqft_above,sqft_basement,sqft_living,sqft_living15,sqft_lot,sqft_lot15,view,waterfront,yr_built,yr_renovated,zipcode

# 3. Data selection

In pandas.DataFrame you could select

1. Row/s by position (integer number \[0 .. number of rows - 1\]) `.iloc` or by DataFrame.index `.loc`:
```
data.loc[0]
data.loc[5:10]
data.iloc[0]
data.iloc[5:10]
```
*Though, this is probably the worst way to manipulate rows.*

2. Columns by name
```
data[columname]
```
3. Row/s and columns
```
data.loc[10, columname]
data.iloc[10, columname]
```
4. Using boolean mask
```
mask = data[columname] > value
data[mask]
```
You could combine multiple conditions using `&` or `|` (and, or)

```
cond1 = data[columname1] > value1
cond2 = data[columname2] > value2
data[cond1 & cond2]
```
5. Using queries `.query()`:
```
value = 5
data.query("columname > value")
```
You could combine multiple conditions using `and`, `or`

```
data.query("(columname1 > value1) and (columname2 > value2)")
```
and others. See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html for more examples.

*Remember to use different quotation marks " or ' for columnname inside a query.*


In [27]:
# setting DataFrame index to be an `id` column, now .loc and .iloc will have different behavior
house_data.index = house_data.id

# dropping `id` column, since now it is an index
house_data.drop('id', axis=1, inplace=True)

# sort data by index for clarity
house_data.sort_index(inplace=True)

In [28]:
house_data

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
1000102     20150422T000000  300000.0         6       3.00         2400   
1000102     20140916T000000  280000.0         6       3.00         2400   
1200019     20140508T000000  647500.0         4       1.75         2060   
1200021     20140811T000000  400000.0         3       1.00         1460   
2800031     20150401T000000  235000.0         3       1.00         1430   
...                     ...       ...       ...        ...          ...   
9842300095  20140725T000000  365000.0         5       2.00         1600   
9842300485  20150311T000000  380000.0         2       1.00         1040   
9842300540  20140624T000000  339000.0         3       1.00         1100   
9895000040  20140703T000000  399900.0         2       1.75         1410   
9900000190  20141030T000000  268950.0         3       1.00         1320   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
1000102         9373     2.0           0     0          3      7        2400   
1000102         9373     2.0           0     0          3      7        2400   
1200019        26036     1.0           0     0          4      8        1160   
1200021        43000     1.0           0     0          3      7        1460   
2800031         7599     1.5           0     0          4      6        1010   
...              ...     ...         ...   ...        ...    ...         ...   
9842300095      4168     1.5           0     0          3      7        1600   
9842300485      7372     1.0           0     0          5      7         840   
9842300540      4128     1.0           0     0          4      7         720   
9895000040      1005     1.5           0     0          3      9         900   
9900000190      8100     1.0           0     0          3      6         880   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
1000102                 0      1991             0    98002  47.3262 -122.214   
1000102                 0      1991             0    98002  47.3262 -122.214   
1200019               900      1947             0    98166  47.4444 -122.351   
1200021                 0      1952             0    98166  47.4434 -122.347   
2800031               420      1930             0    98168  47.4783 -122.265   
...                   ...       ...           ...      ...      ...      ...   
9842300095              0      1927             0    98126  47.5297 -122.381   
9842300485            200      1939             0    98126  47.5285 -122.378   
9842300540            380      1942             0    98126  47.5296 -122.379   
9895000040            510      2011             0    98027  47.5446 -122.018   
9900000190            440      1943             0    98166  47.4697 -122.351   

            sqft_living15  sqft_lot15  
id                                     
1000102              2060        7316  
1000102              2060        7316  
1200019              2590       21891  
1200021              2250       20023  
2800031              1290       10320  
...                   ...         ...  
9842300095           1190        4168  
9842300485           1930        5150  
9842300540           1510        4538  
9895000040           1440        1188  
9900000190           1000        8100  

[21613 rows x 20 columns]

In [29]:
# Select rows by position (int) 

# Q10.1 How many bedrooms have a house on row 777?
# Q10.2 When was built a house on row 9999?
# +Q10.3 How many floors have a house on row 1337?  1
# Q10.4 How many bathrooms have a house on row 314?
# +Q10.5 What is the grade of a house on row 2718?  7

house_data.iloc[1337], house_data.iloc[2718]

(date             20140729T000000
 price                     930000
 bedrooms                       4
 bathrooms                      3
 sqft_living                 2900
 sqft_lot                   10400
 floors                         1
 waterfront                     0
 view                           3
 condition                      5
 grade                          8
 sqft_above                  1530
 sqft_basement               1370
 yr_built                    1959
 yr_renovated                   0
 zipcode                    98008
 lat                      47.5854
 long                    -122.114
 sqft_living15               2820
 sqft_lot15                 10400
 Name: 627300105, dtype: object,
 date             20150326T000000
 price                     537000
 bedrooms                       3
 bathrooms                    2.5
 sqft_living                 1400
 sqft_lot                    4800
 floors                         1
 waterfront                     0
 view          

In [30]:
# Select rows by index (int)

# Q11.1 How many times the house with index 1000102 were sold?
# Q11.2 What is the price of the house with index 9842300095?
# +Q11.3 When was built the house with index 104510440?             1984
# Q11.4 What is the condition of a house with index 252000300?
# +Q11.5 What is the living area (in square feets) of the house with index 1225069038?    13540

house_data.loc[104510440], house_data.loc[1225069038]

(date             20140604T000000
 price                     219950
 bedrooms                       3
 bathrooms                   2.25
 sqft_living                 1500
 sqft_lot                    7615
 floors                         1
 waterfront                     0
 view                           0
 condition                      3
 grade                          7
 sqft_above                  1150
 sqft_basement                350
 yr_built                    1984
 yr_renovated                   0
 zipcode                    98023
 lat                      47.3146
 long                    -122.351
 sqft_living15               1540
 sqft_lot15                  8649
 Name: 104510440, dtype: object,
 date             20140505T000000
 price                   2.28e+06
 bedrooms                       7
 bathrooms                      8
 sqft_living                13540
 sqft_lot                  307752
 floors                         3
 waterfront                     0
 view          

In [31]:
# Using mask or .query syntax select rows/columns (int)

# Q12.1 How many houses were built during American Great Depression (1929–1939)? Including both start and end year.
# +Q12.2 When was built the only house with basement area = 1024 sqft?   2006
# Q12.3 How many houses are with the highest possible grade?
# +Q12.4 When was built a house with maximal number of bedrooms?  1947   
# Q12.5 How many houses were sold for 256000 dollars?

mask = house_data['sqft_basement'] == 1024
house_data[mask]

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
4136980090  20150407T000000  537000.0         4       4.25         4883   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
4136980090     26040     2.0           0     3          3     10        3859   

            sqft_basement  yr_built  yr_renovated  zipcode     lat     long  \
id                                                                            
4136980090           1024      2006             0    98092  47.263 -122.216   

            sqft_living15  sqft_lot15  
id                                     
4136980090           3736        9870

In [32]:
mask = house_data['bedrooms'] == np.max(house_data['bedrooms'])
house_data[mask]

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
2402100895  20140625T000000  640000.0        33       1.75         1620   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
2402100895      6000     1.0           0     0          5      7        1040   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
2402100895            580      1947             0    98103  47.6878 -122.331   

            sqft_living15  sqft_lot15  
id                                     
2402100895           1330        4700

In [33]:
# Using mask or .query syntax select rows/columns (int)

# Q13.1 How many houses with the waterfront (=1) were built duroing Nixon's presidency (1969—1974)? Including both start and end year.
# +Q13.2 How many houses, built before first human in space (<1961), have high condition (=5)?   1153
# +Q13.3 How many houses are with 6 bedrooms and less than 2000 sqft living area?  20
# Q13.4 What was the price of a house with 5 bathrooms, built in 1998 and graded with 10 score?
# Q13.5 How many floors has a house built in 1999 with 5 bedrooms and 3400 sqft living area?
house_data.query("(yr_built < 1961) and (condition == 5)").shape

(1153, 20)

In [34]:
house_data.query("(bedrooms == 6) and (sqft_living < 2000)").shape

(20, 20)

In [35]:
# Using mask or .query syntax select rows/columns and compute simple statistics (float)

# Q14.1 What was the average (sold) price of a houses built in the year of Cuban Missile Crisis (1962)?
# Q14.2 What was the price of the most expensive house sold, built between 1991 and 2000?
# Q14.3 What was the price of the least expensive house sold, built between 1991 and 2000?
# +Q14.4 What is the median number of bathrooms in houses with grade above 9 (10 and more)?  3.25
# +Q14.5 What is the median grade of houses with most popular zipcode value?  7.0
a = house_data.query("(grade > 9)")
a['bathrooms'].median()

3.25

In [36]:
house_data['zipcode'].value_counts()

98103    602
98038    590
98115    583
98052    574
98117    553
        ... 
98102    105
98010    100
98024     81
98148     57
98039     50
Name: zipcode, Length: 70, dtype: int64

In [37]:
b = house_data.query("(zipcode == 98103)")
b['grade'].median()

7.0

# 4. Creating new columns


Creating new column of pandas.DataFrame is as easy as:
```
data['new_awesome_column'] = [] 
```
that's it. But such a column is relatively useless. 
Typically, you would compute something new based on existing data and save it in a new column. 
For example one might want to compute total area of the house as a sum of all `sqft_` columns, or
create a boolean column of whether the house has `grade` > 2 or anything else:

```
data['total_area'] = data[col1] + data[col2] + ...
data['high_value'] = data[col] > 5
```

Pandas also provides another powerfull tool: `.apply`, `.map()`, `.applymap()` methods (they are kinda the same, but not quite). https://stackoverflow.com/questions/19798153/difference-between-map-applymap-and-apply-methods-in-pandas .
They allow you to *apply* some function to every value in the column/s (row-wise) or row (column-wise) or cell (element-wise). For example, same computations of `total_area` and `high_value` using `.apply()`:
```
data['total_area'] = data[[col1, col2, col3]].apply(sum, axis=1)

```
you are not restricted to existent functions, `.apply()` accepts any function (including lambda functions):

```
data['total_area'] = data[[col1, col2, col3]].apply(lambda x: x[0]+x[1]+x[2], axis=1)
```
or ordinary python function (if this it should have complex behaviour):
```
def _sum(x):
    total = 0
    for elem in x:
        total += elem
    return total
    
data['total_area'] = data[[col1, col2, col3]].apply(_sum, axis=1) 
```
Many pandas methods has `axis` parameter `axis=0` refers to rows, `axis=1` refers to columns.

*Warning. You should never use for loops to sum numerical elements from the container.*

---


In [38]:
# Create `was_renovated` column. Bool column (0, 1) indicating whether the house was renovated.

In [39]:
def renov(data):
    if data == 0:
        return 0
    else:
        return 1

In [40]:
house_data['was_renovated'] = house_data['yr_renovated'].apply(renov)
house_data['was_renovated'][house_data['was_renovated'] != 0]

id
3600057       1
9000025       1
16000200      1
31000165      1
46100204      1
             ..
9828200187    1
9828200746    1
9828700200    1
9828701690    1
9829200250    1
Name: was_renovated, Length: 914, dtype: int64

In [41]:
# Create new columns using the old ones (new column in your DataFrame)

# Q15.1 Create a `sqft_tot_area` column (sum of all columns with `sqft_` prefix) using any method above
# Q15.2 Create a new column `sqm_tot_area` using `sqft_tot_area` and the fact that 1 foot = 0.3048 meters
# Q15.3 Create a new column `sqm_aver_floor_area` by dividing total area (in meters) by number of floors
# Q15.4 Create a new column `price_cat` by splitting a `price` into 5 ([1..5]) distinct intervals: 0 < x <=20%,
# 20% < x <= 40%, ... 80% < x <= 100% percentiles. You could use `.quantile()` to compute percentiles.
# Q15.5 Create a new bool column `high_class` it is True if the house has grade >= 9 and condition >= 4

house_data['sqft_tot_area'] = house_data['sqft_living'] + house_data['sqft_lot'] + house_data['sqft_above'] + house_data['sqft_basement']  + house_data['sqft_living15']  + house_data['sqft_lot15']
house_data['sqm_tot_area'] = house_data['sqft_tot_area']*0.3048
house_data['sqm_aver_floor_area'] = house_data['sqm_tot_area']/house_data['floors']

In [42]:
def func_perc(data):
    if data < abc[0]:
        return 1
    if data > abc[0] and data < abc[1]:
        return 2
    if data > abc[1] and data < abc[2]:
        return 3
    if data > abc[2] and data < abc[3]:
        return 4
    if data > abc[3] and data < abc[4]:
        return 5

In [43]:
abc = np.array([house_data['price'].quantile(0.2),house_data['price'].quantile(0.4),house_data['price'].quantile(0.6),house_data['price'].quantile(0.8),house_data['price'].quantile(1)])
abc

array([ 298470.,  399500.,  519000.,  700108., 7700000.])

In [44]:
house_data['price'].mean(), house_data['price'].max()

(540088.1417665294, 7700000.0)

In [45]:
#4
house_data['price_cat'] = house_data['price'].apply(func_perc)
house_data.head(25)

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                      
1000102   20150422T000000  300000.0         6       3.00         2400   
1000102   20140916T000000  280000.0         6       3.00         2400   
1200019   20140508T000000  647500.0         4       1.75         2060   
1200021   20140811T000000  400000.0         3       1.00         1460   
2800031   20150401T000000  235000.0         3       1.00         1430   
3600057   20150319T000000  402500.0         4       2.00         1650   
3600072   20150330T000000  680000.0         4       2.75         2220   
3800008   20150224T000000  178000.0         5       1.50         1990   
5200087   20140709T000000  487000.0         4       2.50         2540   
6200017   20141112T000000  281000.0         3       1.00         1340   
7200080   20141104T000000  239000.0         4       2.00         1980   
7200179   20141016T000000  150000.0         2       1.00          840   
7200179   20150424T000000  175000.0         2       1.00          840   
7400062   20140521T000000  299800.0         2       1.00          790   
7600057   20140805T000000  520000.0         3       2.00         1410   
7600065   20140605T000000  465000.0         3       2.25         1530   
7600125   20141218T000000  630000.0         5       1.00         3020   
7600136   20140718T000000  411000.0         2       2.00         1130   
9000025   20141203T000000  496000.0         2       1.00         1420   
11200070  20140721T000000  570000.0         3       2.50         1530   
11200290  20140609T000000  546000.0         3       2.50         1530   
11200400  20140923T000000  617000.0         3       2.50         1910   
11300120  20140630T000000  635000.0         3       2.50         3350   
11500240  20150428T000000  872750.0         3       2.50         2870   
11500890  20150312T000000  843000.0         3       2.50         3130   

          sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                           
1000102       9373     2.0           0     0          3      7        2400   
1000102       9373     2.0           0     0          3      7        2400   
1200019      26036     1.0           0     0          4      8        1160   
1200021      43000     1.0           0     0          3      7        1460   
2800031       7599     1.5           0     0          4      6        1010   
3600057       3504     1.0           0     0          3      7         760   
3600072       5310     1.0           0     0          5      7        1170   
3800008      18200     1.0           0     0          3      7        1990   
5200087       5001     2.0           0     0          3      9        2540   
6200017      21336     1.5           0     0          4      5        1340   
7200080      10585     1.5           0     0          2      6        1980   
7200179      12750     1.0           0     0          3      6         840   
7200179      12750     1.0           0     0          3      6         840   
7400062       5240     1.0           0     0          4      6         790   
7600057       2700     2.0           0     0          4      7        1410   
7600065       1245     2.0           0     0          3      9        1050   
7600125       4800     2.0           0     0          3      7        3020   
7600136       1148     2.0           0     0          3      9         800   
9000025       4635     2.0           0     0          4      7        1420   
11200070      3296     2.0           0     0          3      8        1530   
11200290      3464     2.0           0     0          3      8        1530   
11200400      4488     2.0           0     0          3      8        1910   
11300120      4007     2.0           0     0          3      8        2550   
11500240     13695     2.0           0     0          3     10        2870   
11500890      8750   

In [46]:
def func_tr_f(data):
    if data['grade'] >=9 and data['condition'] >=4:
        return True
    else:
        return False

In [47]:
# Q15.5 Create a new bool column `high_class` it is True if the house has grade >= 9 and condition >= 4
house_data['high_class'] = house_data[['grade', 'condition']].apply(func_tr_f, axis=1)
house_data

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
1000102     20150422T000000  300000.0         6       3.00         2400   
1000102     20140916T000000  280000.0         6       3.00         2400   
1200019     20140508T000000  647500.0         4       1.75         2060   
1200021     20140811T000000  400000.0         3       1.00         1460   
2800031     20150401T000000  235000.0         3       1.00         1430   
...                     ...       ...       ...        ...          ...   
9842300095  20140725T000000  365000.0         5       2.00         1600   
9842300485  20150311T000000  380000.0         2       1.00         1040   
9842300540  20140624T000000  339000.0         3       1.00         1100   
9895000040  20140703T000000  399900.0         2       1.75         1410   
9900000190  20141030T000000  268950.0         3       1.00         1320   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
1000102         9373     2.0           0     0          3      7        2400   
1000102         9373     2.0           0     0          3      7        2400   
1200019        26036     1.0           0     0          4      8        1160   
1200021        43000     1.0           0     0          3      7        1460   
2800031         7599     1.5           0     0          4      6        1010   
...              ...     ...         ...   ...        ...    ...         ...   
9842300095      4168     1.5           0     0          3      7        1600   
9842300485      7372     1.0           0     0          5      7         840   
9842300540      4128     1.0           0     0          4      7         720   
9895000040      1005     1.5           0     0          3      9         900   
9900000190      8100     1.0           0     0          3      6         880   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
1000102                 0      1991             0    98002  47.3262 -122.214   
1000102                 0      1991             0    98002  47.3262 -122.214   
1200019               900      1947             0    98166  47.4444 -122.351   
1200021                 0      1952             0    98166  47.4434 -122.347   
2800031               420      1930             0    98168  47.4783 -122.265   
...                   ...       ...           ...      ...      ...      ...   
9842300095              0      1927             0    98126  47.5297 -122.381   
9842300485            200      1939             0    98126  47.5285 -122.378   
9842300540            380      1942             0    98126  47.5296 -122.379   
9895000040            510      2011             0    98027  47.5446 -122.018   
9900000190            440      1943             0    98166  47.4697 -122.351   

            sqft_living15  sqft_lot15  was_renovated  sqft_tot_area  \
id                                                                    
1000102              2060        7316              0          23549   
1000102              2060        7316              0          23549   
1200019              2590       21891              0          54637   
1200021              2250       20023              0          68193   
2800031              1290       10320              0          22069   
...                   ...         ...            ...            ...   
9842300095           1190        4168              0          12726   
9842300485           1930        5150              0          16532   
9842300540           1510        4538              0          12376   
9895000040           1440        1188              0           6453   
9900000190           1000        8100              0          19840   

            sqm_tot_area  sqm_aver_floor_

In [48]:
# Using mask or .query syntax select rows/columns (float)

# Q16.1 What is the average price of the house of the high_class(=True)?
# Q16.2 What is the average total_area (in meters) of the house from highest price category?
# Q16.3 What is the maximal number of floors amongst houses with the lowest price category?
# +Q16.4 What is the most frequent zipcode amongst houses with the lowest price category?   98023
# +Q16.5 What is the minimal number of bathrooms in houses with high_class=True?  1.0

mask = house_data['price_cat'] == 1
house_data['zipcode'][mask].value_counts()

98023    325
98042    282
98001    247
98168    220
98038    188
        ... 
98107      1
98029      1
98074      1
98112      1
98007      1
Name: zipcode, Length: 61, dtype: int64

In [49]:
#16.5
mask = house_data['high_class'] == True
house_data['bathrooms'][mask].min()

1.0

# 5. Basic date processing

You figure out that column `date` is to harsh for you, so you decided to convert it to a more plausible format:

- Use pandas method `to_datetime()` to convert the date to a good format.
- Exctract `year`, `month`, `day` and `weekday` from your new date column. Save them to separete columns.
- How many columns has your data now?
- Drop column `date`, remember to set `inplace` parameter to True.


**Hint** for datetime formatted date you could extract the `year` as follow:
```
data.date.dt.year
```

Very often date could be a ridiculously rich feature, sometimes it is holidays that matters, sometimes weekends, sometimes some special days like **black friday**. 

Learn how to work with date in Python!

In [50]:
# Create new columns based on `date` column

# Q17.1 Convert date to datetime format
# Q17.2 Extract and store `year`
# Q17.3 Extract and store `month`
# Q17.4 Extract and store `day`
# Q17.5 Extract and store `weekday`
# Q17.6 Create a new column `house_age_10` - the age of the house in full decades (e.g. 9 year old house - 0, 21 year old house - 2),
# using `yr_built` and 'year' columns
house_data['date'] = pd.to_datetime(house_data['date'])
house_data['year'] = house_data.date.dt.year
house_data['month'] = house_data.date.dt.month
house_data['day'] = house_data.date.dt.day
house_data['weekday'] = house_data.date.dt.weekday
house_data['house_age_10'] = abs(house_data['year'] - house_data['yr_built'])//10

In [51]:
house_data[house_data['house_age_10'] == -1]

Empty DataFrame
Columns: [date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15, was_renovated, sqft_tot_area, sqm_tot_area, sqm_aver_floor_area, price_cat, high_class, year, month, day, weekday, house_age_10]
Index: []

[0 rows x 31 columns]

In [52]:
# Drop column `date`
house_data.drop(['date'], axis=1, inplace = True)
house_data 

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
1000102     300000.0         6       3.00         2400      9373     2.0   
1000102     280000.0         6       3.00         2400      9373     2.0   
1200019     647500.0         4       1.75         2060     26036     1.0   
1200021     400000.0         3       1.00         1460     43000     1.0   
2800031     235000.0         3       1.00         1430      7599     1.5   
...              ...       ...        ...          ...       ...     ...   
9842300095  365000.0         5       2.00         1600      4168     1.5   
9842300485  380000.0         2       1.00         1040      7372     1.0   
9842300540  339000.0         3       1.00         1100      4128     1.0   
9895000040  399900.0         2       1.75         1410      1005     1.5   
9900000190  268950.0         3       1.00         1320      8100     1.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  \
id                                                                          
1000102              0     0          3      7        2400              0   
1000102              0     0          3      7        2400              0   
1200019              0     0          4      8        1160            900   
1200021              0     0          3      7        1460              0   
2800031              0     0          4      6        1010            420   
...                ...   ...        ...    ...         ...            ...   
9842300095           0     0          3      7        1600              0   
9842300485           0     0          5      7         840            200   
9842300540           0     0          4      7         720            380   
9895000040           0     0          3      9         900            510   
9900000190           0     0          3      6         880            440   

            yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
id                                                                             
1000102         1991             0    98002  47.3262 -122.214           2060   
1000102         1991             0    98002  47.3262 -122.214           2060   
1200019         1947             0    98166  47.4444 -122.351           2590   
1200021         1952             0    98166  47.4434 -122.347           2250   
2800031         1930             0    98168  47.4783 -122.265           1290   
...              ...           ...      ...      ...      ...            ...   
9842300095      1927             0    98126  47.5297 -122.381           1190   
9842300485      1939             0    98126  47.5285 -122.378           1930   
9842300540      1942             0    98126  47.5296 -122.379           1510   
9895000040      2011             0    98027  47.5446 -122.018           1440   
9900000190      1943             0    98166  47.4697 -122.351           1000   

            sqft_lot15  was_renovated  sqft_tot_area  sqm_tot_area  \
id                                                                   
1000102           7316              0          23549     7177.7352   
1000102           7316              0          23549     7177.7352   
1200019          21891              0          54637    16653.3576   
1200021          20023              0          68193    20785.2264   
2800031          10320              0          22069     6726.6312   
...                ...            ...            ...           ...   
9842300095        4168              0          12726     3878.8848   
9842300485        5150              0          16532     5038.9536   
9842300540        4538              0          12376     3772.2048   
9895000040        1188              0           6453     1966.8744   
9900000190        8100              0          19840     6047.2320   

            sqm_aver_floor_area  price_cat  high_class  year  month  day

In [53]:
# Find some date related information from the data (int, Sunday has index 0)

# +Q18.1 What is the most popular selling weekday?    1
# Q18.2 What is the most popular selling month?
# +Q18.3 What is the least popular selling weekday?   6
# Q18.4 What is the median age of the house (on a first available sold date)? (float)
# Q18.5 How many houses were sold on America's Independence Day (July, 4)?

house_data['weekday'].value_counts()

1    4715
2    4603
0    4099
3    3994
4    3685
5     287
6     230
Name: weekday, dtype: int64

# 6. Groupby
from the documentation https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html

By “group by” we are referring to a process involving one or more of the following steps:

    - Splitting the data into groups based on some criteria.
    - Applying a function to each group independently.
    - Combining the results into a data structure.
    
---
`.groupby()` is one of the most powerfull tool for feature engineering. Very often it is used to group object with the same categorical characteristics and compute some statistics (e.g. mean, max, etc.) of a their numerical characteric. 

Instead of computing average area of houses with high grade you could compute average areas of the
houses for every grade in a single command:

```
data.groupby('grade')['sqm_tot_area'].mean()
```

You could also make multi-column groups:

```
data.groupby(['weekday','grade'])['price'].min()
```
next, you could compute multiple aggregation functions:
```
data.groupby(['weekday','grade'])['price'].agg([min, max])
```

instead of using built-in functions you could compute custom functions using apply:
```
import numpy as np
data.groupby(['condition','grade'])['bathrooms'].apply(lambda x: np.quantile(x, .5))
```

and the coolest thing now is that you can map the results of groupby back on your DataFrame!
```
gp = data.groupby(['condition'])['bathrooms'].median()
data['gp_feature'] = data['condition'].map(gp)
```
Now, if some house has `condition == 2`, its `gp_feature` will be equal to the median number of 
bathrooms amongst all houses with `condition == 2`.

Read more examples in the documentation https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html

In [54]:
# Create some groupby features

# Q19.1 `price_by_class` groupby `high_class` and compute median `price`.
# Q19.2 `price_by_year` groupby `year` and compute median price.
# +Q19.3 `price_by_weekday` groupby `weekday` and compute median price.
# Q19.4 `area_by_price` groupby `price_cat` and compute average `sqft_living`.
# +Q19.5 `floors_by_age` groupby `floors` and compute average age of a house.

house_data.groupby(['weekday'])['price'].median()

weekday
0    453000.0
1    449950.0
2    453000.0
3    442500.0
4    455000.0
5    475000.0
6    471250.0
Name: price, dtype: float64

In [55]:
gp = house_data.groupby(['weekday'])['price'].median()
house_data['price_by_weekday'] = house_data['weekday'].map(gp)

In [56]:
house_data['house_age'] = house_data['year'] - house_data['yr_built']
gp2 = house_data.groupby(['floors'])['house_age'].mean()
house_data['floors_by_age'] = house_data['floors'].map(gp2)

In [57]:
house_data

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
1000102     300000.0         6       3.00         2400      9373     2.0   
1000102     280000.0         6       3.00         2400      9373     2.0   
1200019     647500.0         4       1.75         2060     26036     1.0   
1200021     400000.0         3       1.00         1460     43000     1.0   
2800031     235000.0         3       1.00         1430      7599     1.5   
...              ...       ...        ...          ...       ...     ...   
9842300095  365000.0         5       2.00         1600      4168     1.5   
9842300485  380000.0         2       1.00         1040      7372     1.0   
9842300540  339000.0         3       1.00         1100      4128     1.0   
9895000040  399900.0         2       1.75         1410      1005     1.5   
9900000190  268950.0         3       1.00         1320      8100     1.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  \
id                                                                          
1000102              0     0          3      7        2400              0   
1000102              0     0          3      7        2400              0   
1200019              0     0          4      8        1160            900   
1200021              0     0          3      7        1460              0   
2800031              0     0          4      6        1010            420   
...                ...   ...        ...    ...         ...            ...   
9842300095           0     0          3      7        1600              0   
9842300485           0     0          5      7         840            200   
9842300540           0     0          4      7         720            380   
9895000040           0     0          3      9         900            510   
9900000190           0     0          3      6         880            440   

            yr_built  yr_renovated  zipcode  ...  sqft_lot15  was_renovated  \
id                                           ...                              
1000102         1991             0    98002  ...        7316              0   
1000102         1991             0    98002  ...        7316              0   
1200019         1947             0    98166  ...       21891              0   
1200021         1952             0    98166  ...       20023              0   
2800031         1930             0    98168  ...       10320              0   
...              ...           ...      ...  ...         ...            ...   
9842300095      1927             0    98126  ...        4168              0   
9842300485      1939             0    98126  ...        5150              0   
9842300540      1942             0    98126  ...        4538              0   
9895000040      2011             0    98027  ...        1188              0   
9900000190      1943             0    98166  ...        8100              0   

            sqft_tot_area  sqm_tot_area  sqm_aver_floor_area  price_cat  \
id                                                                        
1000102             23549     7177.7352            3588.8676        2.0   
1000102             23549     7177.7352            3588.8676        1.0   
1200019             54637    16653.3576           16653.3576        4.0   
1200021             68193    20785.2264           20785.2264        3.0   
2800031             22069     6726.6312            4484.4208        1.0   
...                   ...           ...                  ...        ...   
9842300095          12726     3878.8848            2585.9232        2.0   
9842300485          16532     5038.9536            5038.9536        2.0   
9842300540          12376     3772.2048            3772.2048        2.0   
9895000040           6453     1966.8744            1311.2496        3.0   
9900000190          19840     6047.2320            6047.2320        1.0   

            high_cla

In [58]:
# Create some other groupby features
# for this task check out this answer:
# https://stackoverflow.com/questions/47913343/how-to-groupby-and-map-by-two-columns-pandas-dataframe

# +Q20.1 `n_houses_zipcode` groupby `zipcode` and count number of occurences of every unique zipcode
# Q20.2 `n_houses_yr_built` groupby `yr_built` and count number of houses built in each year
# +Q20.3 `price_by_yr_month_`(median, std) groupby `year`, `month` and compute median and std `price`.
# Q20.4 `price_by_grade_age_`(median, std) groupby `grade`, `house_age` and compute median and std `price`.
# Q20.5 `living_by_cond_`(median, std) groupby `waterfront`, `view`, `condition` and compute median and std `sqft_living`.

house_data['zipcode'].value_counts()

98103    602
98038    590
98115    583
98052    574
98117    553
        ... 
98102    105
98010    100
98024     81
98148     57
98039     50
Name: zipcode, Length: 70, dtype: int64

In [59]:
#Q20.1
gp3 = house_data.groupby(['zipcode'])['zipcode'].count()
house_data['n_houses_zipcode'] = house_data['zipcode'].map(gp3)

In [60]:
house_data

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
id                                                                         
1000102     300000.0         6       3.00         2400      9373     2.0   
1000102     280000.0         6       3.00         2400      9373     2.0   
1200019     647500.0         4       1.75         2060     26036     1.0   
1200021     400000.0         3       1.00         1460     43000     1.0   
2800031     235000.0         3       1.00         1430      7599     1.5   
...              ...       ...        ...          ...       ...     ...   
9842300095  365000.0         5       2.00         1600      4168     1.5   
9842300485  380000.0         2       1.00         1040      7372     1.0   
9842300540  339000.0         3       1.00         1100      4128     1.0   
9895000040  399900.0         2       1.75         1410      1005     1.5   
9900000190  268950.0         3       1.00         1320      8100     1.0   

            waterfront  view  condition  grade  sqft_above  sqft_basement  \
id                                                                          
1000102              0     0          3      7        2400              0   
1000102              0     0          3      7        2400              0   
1200019              0     0          4      8        1160            900   
1200021              0     0          3      7        1460              0   
2800031              0     0          4      6        1010            420   
...                ...   ...        ...    ...         ...            ...   
9842300095           0     0          3      7        1600              0   
9842300485           0     0          5      7         840            200   
9842300540           0     0          4      7         720            380   
9895000040           0     0          3      9         900            510   
9900000190           0     0          3      6         880            440   

            yr_built  yr_renovated  zipcode  ...  was_renovated  \
id                                           ...                  
1000102         1991             0    98002  ...              0   
1000102         1991             0    98002  ...              0   
1200019         1947             0    98166  ...              0   
1200021         1952             0    98166  ...              0   
2800031         1930             0    98168  ...              0   
...              ...           ...      ...  ...            ...   
9842300095      1927             0    98126  ...              0   
9842300485      1939             0    98126  ...              0   
9842300540      1942             0    98126  ...              0   
9895000040      2011             0    98027  ...              0   
9900000190      1943             0    98166  ...              0   

            sqft_tot_area  sqm_tot_area  sqm_aver_floor_area  price_cat  \
id                                                                        
1000102             23549     7177.7352            3588.8676        2.0   
1000102             23549     7177.7352            3588.8676        1.0   
1200019             54637    16653.3576           16653.3576        4.0   
1200021             68193    20785.2264           20785.2264        3.0   
2800031             22069     6726.6312            4484.4208        1.0   
...                   ...           ...                  ...        ...   
9842300095          12726     3878.8848            2585.9232        2.0   
9842300485          16532     5038.9536            5038.9536        2.0   
9842300540          12376     3772.2048            3772.2048        2.0   
9895000040           6453     1966.8744            1311.2496        3.0   
9900000190          19840     6047.2320            6047.2320        1.0   

            high_class  year  month  day  weekday  house_age_10  \
id                                                                
1000102          False  2015      4   22  

In [61]:
# Q20.3 `price_by_yr_month_`(median, std) groupby `year`, `month` and compute median and std `price`.
house_data.groupby(['year', 'month'])['price'].median()

year  month
2014  5        465000.0
      6        465000.0
      7        465000.0
      8        442100.0
      9        450000.0
      10       446900.0
      11       435000.0
      12       432500.0
2015  1        438500.0
      2        425545.0
      3        450000.0
      4        476500.0
      5        455000.0
Name: price, dtype: float64

In [62]:
house_data.groupby(['year', 'month'])['price'].std()

year  month
2014  5        356502.789521
      6        388260.134140
      7        346731.122314
      8        370583.428226
      9        372531.666484
      10       399243.064472
      11       355710.228751
      12       353292.716182
2015  1        366053.601187
      2        325998.284607
      3        355051.196554
      4        369928.879358
      5        414821.964244
Name: price, dtype: float64

# 7. Building a regression model

> You do not need to normalize data for tree models, and for linear/knn models this step is essential.

> Remember, that not all of the features in the table are numeric, some of them might be viewed as categorical.

> You may create or drop **any** features you want, except for the features which use `price` (e.g. average price of a house with 5 bedrooms).

In [63]:
house_data.columns

Index(['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15', 'was_renovated', 'sqft_tot_area',
       'sqm_tot_area', 'sqm_aver_floor_area', 'price_cat', 'high_class',
       'year', 'month', 'day', 'weekday', 'house_age_10', 'price_by_weekday',
       'house_age', 'floors_by_age', 'n_houses_zipcode'],
      dtype='object')

In [64]:
# Q21 Drop all generated features which used price column, e.g. price_by_year, price_cat.
house_data.drop('price_cat', axis=1, inplace=True)
house_data.drop('price_by_weekday', axis=1, inplace=True)

In [65]:
house_data.columns

Index(['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15', 'was_renovated', 'sqft_tot_area',
       'sqm_tot_area', 'sqm_aver_floor_area', 'high_class', 'year', 'month',
       'day', 'weekday', 'house_age_10', 'house_age', 'floors_by_age',
       'n_houses_zipcode'],
      dtype='object')

In [66]:
# Q22 Split your data into train and test parts.
# How many records (rows) do you have in train and test tables? (list of int)? 
# Use sklearn.model_selection.train_test_split with test_size=0.33 and random_state=7

from sklearn.model_selection import train_test_split
num_features = ['bedrooms', 'bathrooms', 'sqft_living', 'floors', 'condition', 'grade', 'house_age_10']
cat_features = ['was_renovated', 'waterfront', 'view']

X = house_data[[*num_features, *cat_features]]
y = house_data['price']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=7)
[X_train.shape[0], X_test.shape[0]]

[14480, 7133]

In [67]:
from sklearn.preprocessing import MinMaxScaler

In [68]:
scaler = MinMaxScaler()
scaler.fit(X_train[num_features])

X_train[num_features] = scaler.transform(X_train[num_features])
X_test[num_features] = scaler.transform(X_test[num_features])

C:\Users\nanne\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
C:\Users\nanne\anaconda3\lib\site-packages\pandas\core\indexing.py:965: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
C:\Users\nanne\anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See th

In [69]:
# Create a predictive regression model of a house price.

# +Q23.1 Use linear regression with l2 regularization (Ridge regression)
# Q23.2 Use decision tree regression
# +Q23.3 Use k nearest neighbours regression

In [70]:
# Q23.1 Use linear regression with l2 regularization (Ridge regression)

from sklearn.linear_model import Ridge
reg = Ridge(fit_intercept=True)
reg.fit(X_train, y_train)
y_pred_train = reg.predict(X_train)
y_pred_test = reg.predict(X_test)

In [71]:
from sklearn.metrics import mean_squared_error
def mse(a, b, prec=2):
    return np.round(mean_squared_error(a, b), prec)

In [72]:
train_error = mse(y_pred_train, y_train)
test_error = mse(y_pred_test, y_test)
weights = np.sum(reg.coef_**2)
train_error, test_error

(49233888233.98, 42442793563.12)

In [73]:
def score(self, X, y, sample_weight=None):
    from sklearn.metrics import r2_score
    y_pred = self.predict(X)
    return r2_score(y, y_pred, sample_weight=sample_weight)

In [74]:
score(reg, X_train, y_train), score(reg, X_test, y_test)

(0.6521506268753429, 0.6493111376417459)

In [75]:
reg.coef_, reg.intercept_

(array([-1167756.30499922,   419543.23834535,  2170392.75015476,
           61241.41076683,    75047.56061768,  1520765.35191902,
          405457.46200433,    20731.9294918 ,   598961.14646604,
           48775.51407622]),
 -804596.1426534076)

In [76]:
# +Q23.3 Use k nearest neighbours regression

from sklearn.neighbors import KNeighborsRegressor
reg2 = KNeighborsRegressor()
reg2.fit(X_train, y_train)
y_pred_train2 = reg2.predict(X_train)
y_pred_test2 = reg2.predict(X_test)

train_error2 = mse(y_pred_train2, y_train)
test_error2 = mse(y_pred_test2, y_test)
train_error2, test_error2

(35348316279.25, 45828238344.27)

In [77]:
score(reg2, X_train, y_train), score(reg2, X_test, y_test)

(0.7502555638036412, 0.6213384789355128)

In [78]:
# Use grid search to select optimal hyperparamters of your models. 

# +Q24.1 Alpha for a ridge regression
# Q24.2 Depth for the tree
# +Q24.3 Number of neighbours for the knn

In [79]:
from sklearn.model_selection import GridSearchCV

In [80]:
clf = Ridge()

grid = {
    'alpha':np.logspace(-6,1,20)
}

gs = GridSearchCV(clf, param_grid=grid)
gs.fit(X_train, y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=None, normalize=False, random_state=None,
                             solver='auto', tol=0.001),
             iid='deprecated', n_jobs=None,
             param_grid={'alpha': array([1.00000000e-06, 2.33572147e-06, 5.45559478e-06, 1.27427499e-05,
       2.97635144e-05, 6.95192796e-05, 1.62377674e-04, 3.79269019e-04,
       8.85866790e-04, 2.06913808e-03, 4.83293024e-03, 1.12883789e-02,
       2.63665090e-02, 6.15848211e-02, 1.43844989e-01, 3.35981829e-01,
       7.84759970e-01, 1.83298071e+00, 4.28133240e+00, 1.00000000e+01])},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [81]:
gs.best_estimator_

Ridge(alpha=0.3359818286283781, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [82]:
gs.best_score_

0.6488150106115238

In [83]:
clf2 = KNeighborsRegressor()

grid2 = {
    'n_neighbors': range(1, 50, 1)
}

gs2 = GridSearchCV(clf2, param_grid=grid2)
gs2.fit(X_train, y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=KNeighborsRegressor(algorithm='auto', leaf_size=30,
                                           metric='minkowski',
                                           metric_params=None, n_jobs=None,
                                           n_neighbors=5, p=2,
                                           weights='uniform'),
             iid='deprecated', n_jobs=None,
             param_grid={'n_neighbors': range(1, 50)}, pre_dispatch='2*n_jobs',
             refit=True, return_train_score=False, scoring=None, verbose=0)

In [84]:
gs2.best_estimator_

KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
                    metric_params=None, n_jobs=None, n_neighbors=7, p=2,
                    weights='uniform')

In [85]:
gs2.best_score_

0.626350370190966

In [86]:
# Compute train and test mean squared error for your best models (list of float).

# +Q25.1 Train, test MSE using linear regression with l2 regularization
# Q25.2 Train, test MSE using decision tree regression
# +Q25.3 Train, test MSE using k nearest neighbours regression

In [87]:
# +Q25.1 Train, test MSE using linear regression with l2 regularization
reg = Ridge(fit_intercept=True, alpha=0.3359818286283781)
reg.fit(X_train, y_train)
y_pred_train = reg.predict(X_train)
y_pred_test = reg.predict(X_test)

train_error = mse(y_pred_train, y_train)
test_error = mse(y_pred_test, y_test)
weights = np.sum(reg.coef_**2)
[train_error, test_error]

[49204509170.82, 42517838563.13]

In [88]:
# +Q25.3 Train, test MSE using k nearest neighbours regression
reg2 = KNeighborsRegressor(n_neighbors=7)
reg2.fit(X_train, y_train)
y_pred_train2 = reg2.predict(X_train)
y_pred_test2 = reg2.predict(X_test)

train_error2 = mse(y_pred_train2, y_train)
test_error2 = mse(y_pred_test2, y_test)
[train_error2, test_error2]

[38574779044.31, 44594660277.79]

In [89]:
# Compute train and test R^2 for your best models (list of float).

# +Q26.1 Train, test R^2 using linear regression with l2 regularization
# Q26.2 Train, test R^2 using decision tree regression
# +Q26.3 Train, test R^2 using k nearest neighbours regression

In [90]:
# +Q26.1 Train, test R^2 using linear regression with l2 regularization
[round(score(reg, X_train, y_train),3), round(score(reg, X_test, y_test),3)]

[0.652, 0.649]

In [91]:
# +Q26.3 Train, test R^2 using k nearest neighbours regression
[round(score(reg2, X_train, y_train),3), round(score(reg2, X_test, y_test),3)]

[0.727, 0.632]

In [92]:
# Q27 Which features have largest (by absolute value) weight in your linear model (top 5 features)? (list of str).

In [93]:
reg.coef_

array([-1306326.96457269,   411598.48149312,  2248292.92744115,
          60513.93816921,    75667.37297225,  1509291.16258806,
         406344.90571978,    19981.33163096,   601664.28790928,
          47626.01664863])

In [94]:
coef = abs(reg.coef_)
coef.sort()
coef

array([  19981.33163096,   47626.01664863,   60513.93816921,
         75667.37297225,  406344.90571978,  411598.48149312,
        601664.28790928, 1306326.96457269, 1509291.16258806,
       2248292.92744115])

In [95]:
['sqft_living', 'grade', 'bedrooms', 'waterfront', 'bathrooms']

['sqft_living', 'grade', 'bedrooms', 'waterfront', 'bathrooms']

In [96]:
X

bedrooms  bathrooms  sqft_living  floors  condition  grade  \
id                                                                       
1000102            6       3.00         2400     2.0          3      7   
1000102            6       3.00         2400     2.0          3      7   
1200019            4       1.75         2060     1.0          4      8   
1200021            3       1.00         1460     1.0          3      7   
2800031            3       1.00         1430     1.5          4      6   
...              ...        ...          ...     ...        ...    ...   
9842300095         5       2.00         1600     1.5          3      7   
9842300485         2       1.00         1040     1.0          5      7   
9842300540         3       1.00         1100     1.0          4      7   
9895000040         2       1.75         1410     1.5          3      9   
9900000190         3       1.00         1320     1.0          3      6   

            house_age_10  was_renovated  waterfront  view  
id                                                         
1000102                2              0           0     0  
1000102                2              0           0     0  
1200019                6              0           0     0  
1200021                6              0           0     0  
2800031                8              0           0     0  
...                  ...            ...         ...   ...  
9842300095             8              0           0     0  
9842300485             7              0           0     0  
9842300540             7              0           0     0  
9895000040             0              0           0     0  
9900000190             7              0           0     0  

[21613 rows x 10 columns]

# Make sure your .ipynb is linearly executable 
# Kernel -> Restart & Run All -> No ERROR cells

In [97]:
# Q28 Save your .ipynb file: Name_Surname_HA1.ipynb, you will be asked to upload it into the google form.